## only need to run the first 3 cells

To get all the needed information.

In [1]:
import pandas as pd
import requests
import time
import pickle
import numpy as np
import folium

In [2]:
#Use only University, Institution and Approved Amount
df = pd.read_csv('P3_GrantExport.csv',sep=';', usecols=[6,7,13]) # delimiter is ;
df = df[(df.University.notnull())&(df["Approved Amount"]!='0.00')]

#Nicht zuteilbar is NA in German
df_unis = df[df['University'] != 'Nicht zuteilbar - NA'] 

In [3]:
dict_unis = pickle.load( open( "dict_unis.p", "rb" ) )
dict_private = pickle.load( open( "dict_private.p", "rb" ) )
dict_nop = pickle.load( open( "dict_nop.p", "rb" ) )

In [6]:
gmapsAPIurl = 'https://maps.googleapis.com/maps/api/geocode/json'
gmapsAPIkey = 'AIzaSyClUHUWJ3ZyRkYN3DbilaEGGcAtgvHKsdY'


In [7]:
#list of unique universities
universities = df_unis.University.unique()

In [8]:
# Returns a dict for all universities we can map using google maps API for the list 'universities'
def get_data(universities):
    state = []
    null_names = []
    names = []
    dictionary = {}
    count_good = 0
    count_bad = 0
    for university in universities:
        time.sleep(0.3)
        university_a = university.split(" -",1)[0]
        url = gmapsAPIurl
        parameters = {'address': university_a,'region':'CH','key':gmapsAPIkey}
        r = requests.get(url, params=parameters)
        status=r.json()['status']
        results = r.json()['results']
        if status=='OK':
            flag=1
            count_good=count_good+1
            names.append(university)
            address=[]
            for result in results:
                for address_component in result['address_components']:
                    if flag==1 and address_component['types']==['administrative_area_level_1', 'political']:
                        address.append(address_component['short_name'])
                        print(university_a,":", address )
                        dictionary[university] = address[0]
                        state.append(address[0])
                        flag=0
                        break

        if status=='ZERO_RESULTS':
            count_bad=count_bad+1
            null_names.append(university)
    
    return(dictionary, state, null_names, names, count_good, count_bad)



In [9]:
#run get_data for list of unique univerities
dict_unis, state_unis, null_names_unis, names_unis, count_good_unis, count_bad_unis = get_data(universities)

Université de Genève : ['GE']
Universität Basel : ['BS']
Université de Fribourg : ['FR']
Universität Zürich : ['ZH']
Université de Lausanne : ['VD']
Universität Bern : ['BE']
ETH Zürich : ['ZH']
Universität St. Gallen : ['SG']
Pädagogische Hochschule Graubünden : ['GR']
EPF Lausanne : ['VD']
Pädagogische Hochschule Zürich : ['ZH']
Universität Luzern : ['LU']
Robert Walser-Stiftung Bern : ['BE']
Paul Scherrer Institut : ['AG']
Pädagogische Hochschule St. Gallen : ['SG']
Physikal.-Meteorolog. Observatorium Davos : ['GR']
AO Research Institute : ['Wien']
Kantonsspital St. Gallen : ['SG']
Institut für Kulturforschung Graubünden : ['GR']
Interkant. Hochschule für Heilpädagogik ZH : ['ZH']
Pädagogische Hochschule Bern : ['BE']
Hochschule Luzern : ['LU']
Haute école pédagogique du canton de Vaud : ['VD']
Haute école pédagogique BE, JU, NE : ['NE']
Pädagogische Hochschule Luzern : ['LU']
Istituto Svizzero di Roma : ['Lazio']
Pädag. Hochschule Tessin (Teilschule SUPSI) : ['TI']
Pädagogische Hoc

In [10]:
#Total number of columns
uni_count = df_unis.count().University

In [11]:
uni_count

48314

In [12]:
#Gets the percentage of rows for which we found the canton
def getPercentFound(df, null_names, uniORinsti):
    t = df.copy()
    t['count'] = t.groupby([uniORinsti])[uniORinsti].transform('count')
    t = t[[ uniORinsti, 'count']]
    t = t.drop_duplicates().sort('count')
    t['Found_boolean'] = t[uniORinsti].apply( lambda x: (0 if x in null_names else 1) )
    t['Found'] = t.apply( lambda row: (row['Found_boolean']*row['count']), axis=1)
    sumis = t.sum()
    return(sumis.Found*100/uni_count)

In [13]:
#Run it for the uni mapping we have now
getPercentFound(df_unis, null_names_unis, 'University')

/home/kirtan/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:6: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)


86.109616260297216

In [14]:
# Private institutions and NOP(Non profit organisations) could be in any canton so we need to look at
# institutions for them

df_institutions = df_unis[((df_unis['Institution'].notnull()) & ((df_unis['University']=='NPO (Biblioth., Museen, Verwalt.) - NPO') | (df_unis['University']=='Firmen/Privatwirtschaft - FP')))]
del df_institutions['Approved Amount']

##Private institutions
df_private = df_institutions[df_institutions['University']==('Firmen/Privatwirtschaft - FP')]
df_private1 = df_private['Institution']
private = df_private1.drop_duplicates()

##NOP
df_nop = df_institutions[df_institutions['University']==('NPO (Biblioth., Museen, Verwalt.) - NPO')]
df_nop1 = df_nop['Institution']
nop = df_nop1.drop_duplicates()


In [15]:
# get data for list of institutions in private
dict_private, state_private, null_names_private, names_private, count_good_private, count_bad_private = get_data(private)

Stöckli A. & J. AG : ['AG']
Département de Physique Université de Fribourg : ['FR']
Groupe Limnoceane Institut de géologie Université de Neuchâtel : ['NE']
INFRAS : ['AP']
Siemens Building Technologies AG : ['AG']
Suiselectra Ingenieurunternehmung AG : ['AG']
Prognos AG : ['AG']
Cellulose Attisholz AG : ['AG']
Corporate Communication Von Roll Management AG : ['AG']
HCP Planen und Beraten für das Gesundheitswesen AG : ['AG']
Health Management Institute : ['Jiangsu Sheng']
MOR Informatik AG : ['AG']
Colenco Power Consulting AG : ['AG']
Planconsult W+B AG : ['AG']
Ernst Basler + Partner AG : ['AG']
Aqua-System AG : ['AG']
Interdisziplinäre Berater- und Forschungsgruppe AG : ['AG']
HOLDERBANK Management und Beratung AG : ['AG']
Aarproject AG : ['AG']
Gruppe CORSO : ['LU']
Gesundheitsdepartement des Kantons Basel-Stadt : ['BS']
Ateliers et Laboratoire C R E P H A R T : ['Grand Casablanca']
Forschungszentrum ABB Schweiz AG : ['AG']
Ciba-Geigy AG Forschungszentrum Marly : ['FR']
FEINTOOL AG :

In [16]:
# get data for list of institutions in nop
dict_nop, state_nop, null_names_nop, names_nop, count_good_nop, count_bad_nop = get_data(nop)

Schweizerisches Nationalmuseum Landesmuseum Zürich : ['ZH']
Historische und Antiquarische Gesellschaft zu Basel : ['BS']
Forum Ost-West : ['HE']
Büro für Archäologie der Stadt Zürich : ['ZH']
Kantonsschule Sargans : ['SG']
Staatsarchiv : ['NDS']
Kuratorium der Helvetia Sacra c/o Staatsarchiv Basel : ['BS']
Archäologische Bodenforschung des Kantons Basel-Stadt : ['BS']
Département des travaux publics du Canton de Vaud : ['VD']
Ecole Suisse d'archéologie en Grèce  Univ. de Lausanne : ['VD']
Kommission des Phonogrammarchivs Universität Zürich : ['ZH']
Staatsarchiv Uri : ['UR']
Tschudi-Kommission der AGGS Staatsarchiv Zürich : ['ZH']
Archives d'Etat de Genève : ['GE']
Bibliothek Kantonsschule Luzern : ['LU']
Musées d'art et d'histoire de Genève : ['GE']
Kantonsschule Baden : ['AG']
Museum des Kantons Thurgau : ['TG']
Pro Helvetia : ['BE']
Liceo di Lugano 2 : ['TI']
Laténium Parc et musée d'archéologie de Neuchâtel : ['NE']
Archives de l'Etat de Fribourg : ['FR']
Gymnase du Bugnon : ['Auver

In [17]:
count_good_nop

146

In [18]:
#Doesn't work perfectly for institutions
getPercentFound(df_nop, null_names_nop, 'Institution')

/home/kirtan/anaconda3/lib/python3.5/site-packages/ipykernel/__main__.py:6: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)


0.93968621931531238

In [19]:
dict_nop

{'Aargauer Kantonsschule Baden': 'AG',
 'Abteilung Biochemie Biozentrum Universität Basel': 'BS',
 'Abteilung Geowissenschaften Naturhistorisches Museum Basel': 'BS',
 'Abteilung Mikrobiologie Biozentrum Universität Basel': 'BS',
 'Abwasserverband Altenrhein': 'SG',
 'Alter Ego': 'Hiroshima-ken',
 'Amt für Archäologie des Kantons Thurgau': 'TG',
 'Amt für Raumentwicklung Baudirektion Kanton Zürich': 'ZH',
 'Appellationsgericht Basel-Stadt': 'BS',
 'Arbeitslehrerinnenseminar des Kantons Zürich': 'ZH',
 'Arche Verlag AG': 'AG',
 "Archives D'Etat de Genève Etat de Genève": 'GE',
 "Archives d'Etat de Genève": 'GE',
 "Archives de l'Etat de Fribourg": 'FR',
 "Archives de l'Etat de Neuchâtel": 'NE',
 'Archivio di Stato del Canton Ticino': 'TI',
 'Archäologie Schweiz': 'BE',
 'Archäologische Bodenforschung des Kantons Basel-Stadt': 'BS',
 'Archäologischer Dienst des Kantons Bern': 'BE',
 'BAK Konjunkturforschung Basel AG': 'AG',
 'Bibliothek Kantonsschule Luzern': 'LU',
 'Bibliothek Staatsarch

In [20]:
dict_unis

{'AO Research Institute - AORI': 'Wien',
 'Biotechnologie Institut Thurgau - BITG': 'TG',
 'EPF Lausanne - EPFL': 'VD',
 'ETH Zürich - ETHZ': 'ZH',
 'Facoltà di Teologia di Lugano - FTL': 'TI',
 'Haute école pédagogique BE, JU, NE - HEPBEJUNE': 'NE',
 'Haute école pédagogique du canton de Vaud - HEPL': 'VD',
 'Hochschule Luzern - HSLU': 'LU',
 'Institut für Kulturforschung Graubünden - IKG': 'GR',
 'Instituto Ricerche Solari Locarno - IRSOL': 'TI',
 'Interkant. Hochschule für Heilpädagogik ZH - HfH': 'ZH',
 'Istituto Svizzero di Roma - ISR': 'Lazio',
 'Kantonsspital St. Gallen - KSPSG': 'SG',
 'Paul Scherrer Institut - PSI': 'AG',
 'Physikal.-Meteorolog. Observatorium Davos - PMOD': 'GR',
 'Pädag. Hochschule Tessin (Teilschule SUPSI) - ASP': 'TI',
 'Pädagogische Hochschule Bern - PHBern': 'BE',
 'Pädagogische Hochschule Graubünden - PHGR': 'GR',
 'Pädagogische Hochschule Luzern - PHLU': 'LU',
 'Pädagogische Hochschule Schaffhausen - PHSH': 'SH',
 'Pädagogische Hochschule Schwyz - PHSZ'

In [21]:
dict_private

{'ABW Architektur + Raumplanung AG': 'AG',
 'ASIT': 'Van',
 'Aarproject AG': 'AG',
 'Abteilung Pharmakologie/Neurobiologie Biozentrum der Universität Basel': 'BS',
 'Albrecht & Partner AG': 'AG',
 'Alcatel Schweiz AG': 'AG',
 'Alter Ego': 'Hiroshima-ken',
 'Aqua-System AG': 'AG',
 'Ateliers et Laboratoire C R E P H A R T': 'Grand Casablanca',
 'B & S Ingenieur AG': 'DL',
 'BMG Engineering AG': 'AG',
 'BSS Volkswirtschaftliche Beratung AG': 'AG',
 'Batrec AG': 'AG',
 'Bioengineering AG': 'AG',
 'Biospectra AG': 'AG',
 'Büro für arbeits- und sozialpolitische Studien (Büro BASS), Bern': 'BE',
 "CASIN - Centre for applied studies in international negociations / centre d'étude pratiques de la...": 'BC',
 'CTU - Conzepte Technik Umwelt AG': 'Sichuan Sheng',
 'Carbotech AG': 'AG',
 'Cellulose Attisholz AG': 'AG',
 'Ciba-Geigy AG Forschungszentrum Marly': 'FR',
 'Colenco Power Consulting AG': 'AG',
 'Corporate Communication Von Roll Management AG': 'AG',
 'Departement Geschichte Universität Ba

In [22]:
# Save a dictionary into a pickle file.
# Don't need to run this again!

# pickle.dump( dict_unis, open( "dict_unis.p", "wb" ) )
# pickle.dump( dict_private, open( "dict_private.p", "wb" ) )
# pickle.dump( dict_nop, open( "dict_nop.p", "wb" ) )

## Stuff Left to do: 
* DONE Planning (10 min)
* Read Documentation for how to use map / Folium (25 min)
* Put the dictionary into a column (15 min + ) 
  * university
  * private
  * nop
* Manually add missing ones to get to 95% (10 min)
  * 95% of what?? universities or nop?
* Clean dictionary column for canton (15 min)
  * account for missing values, long values
* Map the DataFrame into the Chloropeth (15 min)
* Bonus -- difference b/w areas Röstigraben (?)

## Mapping Dictionary into DataFrame Column

### After running the first the three first cells, run all the following cells:

In [4]:
df_unis

Institution  \
1      Faculté de Psychologie et des Sciences de l'Ed...   
2      Kommission für das Corpus philosophorum medii ...   
3      Abt. Handschriften und Alte Drucke Bibliothek ...   
4                           Schweiz. Thesauruskommission   
5      Séminaire de politique économique, d'économie ...   
6      Institut für ökumenische Studien Université de...   
7              Ostasiatisches Seminar Universität Zürich   
8                                                    NaN   
9      Laboratoire de Didactique et Epistémologie des...   
10     Klinische Psychologie und Psychotherapie Insti...   
11     Schweizerische Rechtsquellen c/o Universität Z...   
12     Département de Sociologie Faculté des Sciences...   
13     Institut für Volkswirtschaft WWZ / FGS Univers...   
14            Soziologisches Institut Universität Zürich   
15                                 Forschungszentrum FAL   
16     Faculté de Psychologie et des Sciences de l'Ed...   
17            Soziologisches Institut Universität Zürich   
19                                                   NaN   
20     Faculté de Psychologie et des Sciences de l'Ed...   
21                                                   NaN   
22     Forschungsbereich Politisches Verhalten und Oe...   
23     Département des langues et des littératures fr...   
26            Schweizerische Gesellschaft für Volkskunde   
27                         Kuratorium Carl J. Burckhardt   
28                                                   NaN   
29     Eidg. Forschungsanstalt für Wald, Schnee und L...   
30     Schweizerisches Nationalmuseum Landesmuseum Zü...   
31     Fondation Hardt pour l'étude de l'Antiquité cl...   
32        Institut für Iberoromanistik Universität Basel   
33     Historische und Antiquarische Gesellschaft zu ...   
...                                                  ...   
63900      Global Studies Institute Université de Genève   
63902  Istituto di Studi Italiani Facoltà di Scienze ...   
63903  Institut Ethique Histoire Humanités - iEH2 Fac...   
63904  Institut d'Archéologie et des Sciences de l'An...   
63910  Pädagogische Hochschule Fachhochschule Nordwes...   
63913    Seminar für Filmwissenschaft Universität Zürich   
63915                Deutsches Seminar Universität Basel   
63916               Englisches Seminar Universität Basel   
63917                     Centro Stefano Franscini (CSF)   
63918             Romanisches Seminar Universität Zürich   
63919             Romanisches Seminar Universität Zürich   
63920  Institut d'études politiques, historiques et i...   
63921           Theologische Fakultät Universität Zürich   
63922  Institut des Sciences sociales du Politique IS...   
63927  Département d'Informatique Université de Fribourg   
63928         Institut für Tieranatomie Universität Bern   
63932  Institut für Germanistik Philosophisch-histori...   
63934  Centre interfacultaire en droits de l'enfant  ...   
63935                         Haute Ecole Pédagogique VD   
63939  Institut Sozialmanagement und Sozialpolitik Ho...   
63950  Fondation Jean Monnet pour l'Europe Centre de ...   
63951  Planning of Landscape and Urban Systems Instit...   
63952  Forschungsanstalt Agroscope Reckenholz-Tänikon...   
63953  Forschung und Beratung in Wirtschaft und Polit...   
63955  Zentrum für Religion, Wirtschaft und Politik T...   
63958         Abteilung für Indologie Universität Zürich   
63959  Institut Religions, Cultures, Modernité FTSR U...   
63963     Haute École d'art et de design - Genève HES-SO   
63964  Departement Design Zürcher Hochschule der Küns...   
63965  Institute Performing Arts and Film Zürcher Hoc...   

                                              University Approved Amount  
1                              Université de Genève - GE        41022.00  
2                NPO (Biblioth., Museen, Verwalt.) - NPO        79732.00  
3                                 Universität Basel - BS        52627.00  
4                NPO (Bi

In [5]:
# mapping University and Insitutions to cantons
df_unis["Canton"] = df_unis["University"].map(dict_unis)
df_unis.ix[df_unis.University ==('Firmen/Privatwirtschaft - FP'), 'Canton'] = df_unis[df_unis["University"] ==('Firmen/Privatwirtschaft - FP')]['Institution'].map(dict_private)
df_unis.ix[df_unis.University ==('NPO (Biblioth., Museen, Verwalt.) - NPO'), 'Canton'] = df_unis[df_unis["University"] ==('NPO (Biblioth., Museen, Verwalt.) - NPO')]['Institution'].map(dict_nop)

/Users/lemr/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  from ipykernel import kernelapp as app
/Users/lemr/anaconda/lib/python3.5/site-packages/pandas/core/indexing.py:132: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)
/Users/lemr/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-v

In [6]:
df_unis[pd.isnull(df_unis['Canton'])].shape #6713 null

(6082, 4)

In [7]:
df_unis[pd.isnull(df_unis['Canton'])].shape #6713 null
# 6535 null
# 6089
df_unis.shape

print("Percentage Filled",(48314-6089)/48314)

Percentage Filled 0.8739702777662789


In [8]:
df_unis.Canton.unique()

array(['GE', nan, 'BS', 'FR', 'ZH', 'VD', 'BE', 'HE', 'SG', 'NDS', 'GR',
       'UR', 'LU', 'AG', 'TG', 'TI', 'NE', 'Auvergne-Rhône-Alpes',
       'Bourgogne Franche-Comté', 'ZG', 'Pichincha', 'SH',
       'Languedoc-Roussillon-Midi-Pyrénées', 'VS', 'NRW', 'GP', 'Wien',
       'AP', 'Jiangsu Sheng', 'Grand Casablanca', 'WB', 'SO',
       'Mie Prefecture', 'England', 'Nordwest', 'Hiroshima-ken',
       'Kareliya Republits', 'Guangdong', 'BC', 'Sichuan Sheng', 'DL',
       'Van', 'OW', 'Shida Kartli', 'Saitama-ken', 'BL', 'Lazio', 'Edo',
       'SZ', 'Chocó'], dtype=object)

In [9]:
df_unis_canton=df_unis.copy()

In [10]:
#Keep only the swiss cantons (I do not know if we can do this step with a dictionary will all the abreviations of the cantons)
df_unis_canton=df_unis_canton[(df_unis_canton.Canton.str[0]>=chr(65)) & (df_unis_canton.Canton.str[0]<=chr(90)) & 
                              (df_unis_canton.Canton.str[1]>=chr(65)) & (df_unis_canton.Canton.str[1]<=chr(90)) & 
                              (df_unis_canton.Canton!='NDS') & (df_unis_canton.Canton!='NRW') & (df_unis_canton.Canton!='AP') &
                              (df_unis_canton.Canton!='BC') & (df_unis_canton.Canton!='CA') & (df_unis_canton.Canton!='DL') &
                              (df_unis_canton.Canton!='HE') & (df_unis_canton.Canton!='WB')]

In [11]:
df_unis_canton.sort('Canton').Canton.unique()

/Users/lemr/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:1: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)
  if __name__ == '__main__':


array(['AG', 'BE', 'BL', 'BS', 'FR', 'GE', 'GP', 'GR', 'LU', 'NE', 'OW',
       'SG', 'SH', 'SO', 'SZ', 'TG', 'TI', 'UR', 'VD', 'VS', 'ZG', 'ZH'], dtype=object)

## Task 2: Building the Choropleth Map

Swiss maps with the cantons using a DataFrame with Folium
correct "key_on" value

Fix: AttributeError: 'NoneType' object has no attribute 'get'

**To Use DataFrame for Map**

To use a Dataframe (in the DF, I used the id as the _Cantons_ and I put random values inside the _Values_ field), you need to do this:
```python
swiss_map.choropleth(geo_path=cantons_geo, 
                     data=df,
                     columns=['Cantons', 'Values'],
                     key_on='feature.id',
                     topojson='objects.cantons',
                     fill_color='YlGn',
                     legend_name = 'Random numbers'
                    )
```

In you DataFrame, you must be sure that *all the id present in the topojson are also present in the column of the DF*. If it's not the case, just add the missing ids..

In [12]:
df_unis_canton['Approved Amount'] = pd.to_numeric(df_unis_canton['Approved Amount'], errors='coerce')

In [13]:
df_unis_canton = pd.DataFrame(df_unis_canton.groupby('Canton').sum()['Approved Amount'].apply(np.round))

In [14]:
df_unis_canton.reset_index(inplace=1) # what I've been looking for

In [15]:
df_unis_canton.columns = ['Canton', 'Amount']

In [16]:
df_unis_canton.head()

Canton        Amount
0     AG  1.450655e+08
1     BE  1.543377e+09
2     BL  2.571390e+05
3     BS  1.367958e+09
4     FR  4.603587e+08

## Adding Missing Cantons

In [24]:
missing_cantons = {'Canton' : ['NW', 'GL', 'AR', 'AI', 'JU'],
                   'Amount' : np.ones(5)}

In [25]:
missing_cantons = pd.DataFrame(missing_cantons)

In [26]:
missing_cantons

Amount Canton
0     1.0     NW
1     1.0     GL
2     1.0     AR
3     1.0     AI
4     1.0     JU

In [27]:
df_unis_canton_total = pd.concat([df_unis_canton,missing_cantons])

In [28]:
# DataFrame with Canton and Total Funding
df_unis_canton_total.sort('Amount')
df_unis_canton_total['Amount'] = np.log10(df_unis_canton_total['Amount']) 
df_unis_canton_total.fillna(0)

/Users/lemr/anaconda/lib/python3.5/site-packages/ipykernel/__main__.py:2: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)
  from ipykernel import kernelapp as app


Amount Canton
0   8.161564     AG
1   9.188472     BE
2   5.410168     BL
3   9.136073     BS
4   8.663096     FR
5   9.266798     GE
6   5.576836     GP
7   7.181358     GR
8   7.733446     LU
9   6.458570     NE
10  5.140181     OW
11  7.949418     SG
12  5.973850     SH
13  5.941711     SO
14  5.971531     SZ
15  6.808672     TG
16  6.600956     TI
17  5.021189     UR
18  9.374811     VD
19  5.856810     VS
20  6.033992     ZG
21  9.542195     ZH
0   0.000000     NW
1   0.000000     GL
2   0.000000     AR
3   0.000000     AI
4   0.000000     JU

In [76]:
import branca.colormap as cm

colormap = linear.YlGn.scale(
   0,
   10)

colormap = colormap.to_step(10)
colormap

In [77]:
import json

geo_json_data = json.load(open(r'ch-cantons.geojson.json'))

In [78]:
canton_dict1 = df_unis_canton_total.set_index('Canton')['Amount']

In [81]:
m = folium.Map(location=[47, 8],tiles='cartodbpositron', zoom_start=8)

folium.GeoJson(
    geo_json_data,
    style_function=lambda feature: {
        'fillColor': colormap(canton_dict1[feature['id']]),
        'color': 'black',
        'weight': 2,
        'dashArray': '1',
        'fillOpacity': 0.7,
    }
).add_to(m)

In [82]:
colormap.caption = 'Log of total amount'
colormap.add_to(m)

m

cantons_geo = r'ch-cantons.topojson.json'

swiss_map = folium.Map(location=[47, 8],tiles='cartodbpositron', zoom_start=7.7)
swiss_map.geo_json(geo_path=cantons_geo, 
                     data=df_unis_canton_total,
                     columns=['Canton', 'Amount'],
                     key_on='feature.id',
                     topojson='objects.cantons',
                     fill_color='YlGn', fill_opacity=0.75, line_opacity=0.2,
                     legend_name = 'Funding Amount',
                     threshold_scale=[0, 5, 6, 7, 8, 9] #Scale depends on the maximum and minimum value in "Amount"
                    )
swiss_map.save('swiss_cantons.html')


swiss_map

## Task 3: Bonus
BONUS: using the map you have just built, and the geographical information contained in it, could you give a rough estimate of the difference in research funding between the areas divided by the Röstigraben?

-- needs Task 1 and Task 2 to be completed, should be less work than task 1 and 2. Probably statistical test.